### ISSUE TITLE
We should publish a `v1` tag for the databuilder docker image

### ISSUE THREAD

**Issue by evansd** [opensafely-core/databuilder#564]
Initially, this tag would be identical to the `latest` tag. But it gives us the option to break backwards compatibility by freezing `v1` and starting to release to `v2`. The key thing is that all projects should use `databuilder:v1` as the image reference rather than `databuilder:latest` (which should be hard-banned) and that will give us the future flexibility we need.

This is the same approach we used for the [research-action](https://github.com/opensafely-core/research-action) and it served us well there.

Exactly how we apply the same technique using docker tags rather than git tags is still an open question I think.

**Comment by StevenMaude** [comment]
To clarify, we deliberately don't push a `latest` tag for the `databuilder` image to simplify its versioning:

```bash
$ docker pull ghcr.io/opensafely-core/databuilder:latest
Error response from daemon: manifest unknown
```

We made this change when we moved to publishing `databuilder` images instead of just `cohortextractor-v2`. `cohortextractor-v2` does have a now "old" latest tag, but [we'll be rid of that image soon anyway](https://github.com/opensafely-core/job-runner/issues/448).

But, slightly confusingly, [some of the tests in various bits of our code **do** refer to `databuilder:latest`](https://cs.github.com/?scopeName=All+repos&scope=&q=org%3Aopensafely-core+%22databuilder%3Alatest%22), so perhaps these should be changed to reflect actual usage.

**Comment by StevenMaude** [comment]
(I'd forgotten that the code search preview wasn't just available for everyone globally, but, yes, that was the search I was going for.)

**Comment by rebkwok** [comment]
> To clarify, we deliberately don't push a latest tag for the databuilder image to simplify its versioning

Does this mean we actually _don't_ want to tag the image as v1?

**Comment by evansd** [comment]
> Does this mean we actually don't want to tag the image as v1?

I might not be quite following here. I thought the idea was that we have a tag called `v1` which, for the moment, behaves exactly like `latest` does on cohortextractor. i.e. it always points to the most recent build of the image. But unlike `latest` this gives us the option further down the road to stop updating the `v1` tag and start publishing a `v2` tag instead —which means we can safely break backwards compatibility.

**Comment by StevenMaude** [comment]
> I thought the idea was that we have a tag called `v1` which, for the moment, behaves exactly like `latest` does on cohortextractor. i.e. it always points to the most recent build of the image. But unlike `latest` this gives us the option further down the road to stop updating the `v1` tag and start publishing a `v2` tag instead —which means we can safely break backwards compatibility.

Yes, this is what was intended :+1:

***

I'll mention here, even if everyone knows, because it's useful context for why we're avoiding `latest` and it's not an obvious behaviour.
 
**Docker uses `latest` as a tag when pushing an *untagged* image.**

:warning: The confusion is that there are no guarantees that a Docker image with the `latest` tag is the latest image version. Let's say I push a tagged image and then push an untagged version of the same image. Then the untagged version will be older, but still have the `latest` tag.

(If we ever did want a tag always pointing to the latest image without specifying a version, we can also have that. But we should use a tag other than `latest`.)

**Comment by StevenMaude** [comment]
I think the only tasks here are:

- [ ] Decide when Data Builder's design has stabilised enough to have a `v1`.
  * This might be when we're no longer intending to make frequent breaking changes, so we're not having to frequently bump the image version in any study code to track recent updates.
  * It might also be when there's some minimal functionality in Data Builder. Maybe when #470 is complete?
  * Or a combination of both these aspects.
- [ ] Ensure the major version tag is incremented.
  * The Git tag should get bumped to a new major version when [`BREAKING CHANGE:` is written in a commit message](https://github.com/mathieudutour/github-tag-action#bumping).
  * The Docker tag for the pushed image should then automatically match the Git tag.

Feel free to add or correct me, if I'm wrong.

**Comment by evansd** [comment]
Do we want to start by publishing a `v0`, just so we can test things end-to-end without having to wait for API stability?

**Comment by StevenMaude** [comment]
@evansd: There are [images under the opensafely-core organisation](https://github.com/opensafely-core/base-docker/pkgs/container/databuilder), but linked to the `base-docker` image (so I think that's why they don't show in packages here).

Maybe I've missed something else though?

![databuilder-images](https://user-images.githubusercontent.com/3818079/178478419-a59412ec-bf0c-4a05-9019-bcbd0dfafbeb.png)

**Comment by evansd** [comment]
Oh, you're quite right. That's weird. I wonder how we get them linked to the right repo?

**Comment by evansd** [comment]
The above link seems to only be relevant for packages which aren't linked to a repository at all, rather than ones linked to the wrong repository. But I think I know what's happening here. 

The base-docker Dockerfile contains set the `org.opencontainers.image.source` label to point to the base-docker repo:
https://github.com/opensafely-core/base-docker/blob/8d9de9f1c46816eca95b28f15e6aecc128f90b3e/Dockerfile#L12

So hopefully all we need to do is set that label in the databuilder Dockerfile and it will all just work :crossed_fingers:

**Comment by StevenMaude** [comment]
That's a good observation. I opened #623 to add the correct Docker image labels.

**Comment by evansd** [comment]
@StevenMaude Now that #623 is merged I think your [earlier comment](https://github.com/opensafely-core/databuilder/issues/564#issuecomment-1181618292) stands and there's nothing specific left to do on this ticket and it can be closed.

We'll automatically publish v1 once we issue our first breaking change, and deciding at exactly what point to do that feels like a separate issue.

I'll leave it for you to close though in case you disagree.

**Comment by StevenMaude** [comment]
Yep, agree :+1: — this will happen naturally when we decide to do so.
